### PR TITLE
fix: Add edge case handling to traffic routing

### PR DESCRIPTION
**Pull request by zezaeoh**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: rollout/trafficrouting.go [hunk 1]
```diff
@@ -98,6 +98,10 @@ func (c *rolloutContext) reconcileTrafficRouting() error {
 	desiredWeight := int32(0)
 	if c.rollout.Status.StableRS == c.rollout.Status.CurrentPodHash {
 		// when we are fully promoted. desired canary weight should be 0
+	} else if c.pauseContext.IsAborted() {
+		// when promote aborted. desired canary weight should be 0
+	} else if c.newRS.Status.AvailableReplicas == 0 {
+		// when newRS available replicas num is 0. never weight to canary
```
Comment (jessesuen) [inline rollout/trafficrouting.go:104]: This change is significant, and a surprising oversight if we are not handling this correctly. I will take some time to understand the implications of the change.

**Comment by jessesuen** [comment]
Looks like we have a few tests (both unit and e2e) that are asserting the opposite of this change. So at the very least the existing tests would need to change.

But we would also want new tests to verify this change. This is assuming the change is correct -- I still need to play around and what the behavior should be vs. what is actually happening.

**Comment by zezaeoh** [comment]
Hi, @jessesuen 
Thanks a lot for your review.

I would like to cover next cases with this changes

1. abort case
```go
	} else if c.pauseContext.IsAborted() {		
		// when promote aborted. desired canary weight should be 0	
	} ...
``` 
---
when promotion is aborted, controller scale down replicas before change canary weight to 0
because of this logic
```go
                atDesiredReplicaCount := replicasetutil.AtDesiredReplicaCountsForCanary(c.rollout, c.newRS, c.stableRS, c.otherRSs)
		if !atDesiredReplicaCount {
			// Use the previous weight since the new RS is not ready for a new weight
			for i := *index - 1; i >= 0; i-- {
				step := c.rollout.Spec.Strategy.Canary.Steps[i]
				if step.SetWeight != nil {
					desiredWeight = *step.SetWeight
					break
				}
			}
		} ...
```
this behavior leads some request drops 

2. when promotion aborted and then start `promotefull` action
```go
	} else if c.newRS == nil || c.newRS.Status.AvailableReplicas == 0 {
		// when newRS available replicas num is 0. never weight to canary
	} ...
```
---
Suppose we have a rollout resource with a canary strategy like this:
```yaml
    canary:
      steps:
        - setWeight: 20
        - pause: {}
        - setWeight: 40
        - pause: {}
        - setWeight: 60
        - pause: {}
        - setWeight: 100
        - pause: {}
```

when promotion is aborted, then canary service's endpoint will be empty.
If `promotefull` action is executed in this situation,
despite the fact that newRS has just been created, trafficrouting sets the canary weight to 100

this behavior leads request drops

**Comment by jessesuen** [comment]
Thanks for the explanation. The change makes sense now and looks like all tests are passing.

**Review by jessesuen — APPROVED** [review]

**Pull request merged**